### ISSUE TITLE
rr pack takes 33 minutes to run - many mmap_clone files

### ISSUE THREAD

**Issue by karlt** [rr-debugger/rr#3529]
A recent invocation of `rr pack` with a modest duration recording with an optimized build of Firefox took 33 minutes to run.  This was pernosco trace VsonywfDAvU.  There were 88 mmap_clone_*_libxul.so files in the trace. This was with a recent rr ba2f43265637bafcbbd1eda9fc8cd4a8acd05395 5.60, but I was getting similar behavior before pulling 25e6edc5..ba2f4326.  The trace was recorded with `-n` as a somewhat effective workaround for https://github.com/rr-debugger/rr/issues/3444 and https://github.com/rr-debugger/rr/issues/3468

Usually I use rr with a debug build of Firefox and `rr pack` takes around 5 to 8 minutes to run, where there are 20 mmap_clone_*_libxul.so files.

**Comment by rocallahan** [comment]
Was there an unusually large number of processes in this run? That's the main reason I would expect to see lots of libxul clones.

**Comment by karlt** [comment]
10 firefox executions, but several forks. 26 processes in total < 88.
```
% rr ps
PID     PPID    EXIT    CMD
674588  --      0       obj-opt/dist/bin/firefox --profile /tmp/fx-profile --allow-downgrade
674590 (1)      674588  0       (forked without exec)
674601  674588  0       /var/karl/moz/obj-opt/dist/bin/glxtest -f 13
674612  674588  0       /usr/bin/lsb_release -idrc
674631  674612  4       getopt -T
674632  674612  0       getopt -o acdhirsvp -l all,codename,description,help,id,release,short,version,program_version -n lsb_release -- -idrc
674633  674612  0       (forked without exec)
674634  674633  0       (forked without exec)
674635  674633  0       tr -d [:blank:]
674649  674588  0       /var/karl/moz/obj-opt/dist/bin/firefox -contentproc -parentBuildID 20230524175546 -prefsLen 25974 -prefMapSize 233593 -appDir /var/karl/moz/obj-opt/dist/bin/browser {70cf3528-0194-4395-9d86-bb09d4217ab7} 674588 socket
674651  674649  0       (forked without exec)
674820  674588  0       /var/karl/moz/obj-opt/dist/bin/firefox -contentproc -childID 1 -isForBrowser -prefsLen 26219 -prefMapSize 233593 -jsInitLen 243676 -parentBuildID 20230524175546 -appDir /var/karl/moz/obj-opt/dist/bin/browser {5fd9718d-075a-448e-bc4e-7c6790499655} 674588 tab
674821  674820  0       (forked without exec)
675343  674588  0       /var/karl/moz/obj-opt/dist/bin/firefox -contentproc -childID 2 -isForBrowser -prefsLen 28062 -prefMapSize 233593 -jsInitLen 243676 -parentBuildID 20230524175546 -appDir /var/karl/moz/obj-opt/dist/bin/browser {605d8afb-03a2-4fd5-84c5-aa5c6fae3d4b} 674588 tab
675344  675343  0       (forked without exec)
675351  674588  0       /var/karl/moz/obj-opt/dist/bin/firefox -contentproc -childID 3 -isForBrowser -prefsLen 30827 -prefMapSize 233593 -jsInitLen 243676 -parentBuildID 20230524175546 -appDir /var/karl/moz/obj-opt/dist/bin/browser {06309e60-8523-4d79-a17a-253c51f67d3a} 674588 tab
675353  675351  0       (forked without exec)
675490  674588  0       /var/karl/moz/obj-opt/dist/bin/firefox -contentproc -parentBuildID 20230524175546 -sandboxingKind 0 -prefsLen 31771 -prefMapSize 233593 -appDir /var/karl/moz/obj-opt/dist/bin/browser {7d84f4b5-b914-44d2-90e2-1adf3167fa0e} 674588 utility
675494  674588  0       /var/karl/moz/obj-opt/dist/bin/firefox -contentproc -childID 4 -isForBrowser -prefsLen 29006 -prefMapSize 233593 -jsInitLen 243676 -parentBuildID 20230524175546 -appDir /var/karl/moz/obj-opt/dist/bin/browser {d3e595b9-419f-4923-953e-821c005b2cf6} 674588 tab
675495  674588  0       /var/karl/moz/obj-opt/dist/bin/firefox -contentproc -childID 5 -isForBrowser -prefsLen 29006 -prefMapSize 233593 -jsInitLen 243676 -parentBuildID 20230524175546 -appDir /var/karl/moz/obj-opt/dist/bin/browser {07e3238e-9dba-4a4d-9cdc-7d8194e9c8af} 674588 tab
675498  675494  0       (forked without exec)
675500  675495  0       (forked without exec)
675503  674588  0       /var/karl/moz/obj-opt/dist/bin/firefox -contentproc -childID 6 -isForBrowser -prefsLen 29006 -prefMapSize 233593 -jsInitLen 243676 -parentBuildID 20230524175546 -appDir /var/karl/moz/obj-opt/dist/bin/browser {963f6f8d-05e6-468f-85c8-060f32d85a9c} 674588 tab
675505  675503  0       (forked without exec)
675815  674588  0       /var/karl/moz/obj-opt/dist/bin/firefox -contentproc -parentBuildID 20230524175546 -prefsLen 31771 -prefMapSize 233593 -appDir /var/karl/moz/obj-opt/dist/bin/browser {457c0c5e-3d1b-4d57-9eab-65c593293076} 674588 rdd
675821  675815  0       (forked without exec)
```
This is similar to `rr ps` from an older trace with only 20 clones.

**Comment by khuey** [comment]
There's only one libxul.so in the uploaded trace, so packing is working correctly.

Does btrfs not have a "are these two fds reflinked to the same data?" ioctl, and we end up hashing everything?

**Comment by rocallahan** [comment]
I think we could probably accelerate this using the FIEMAP ioctl.

**Comment by rocallahan** [comment]
https://www.kernel.org/doc/Documentation/filesystems/fiemap.txt

**Comment by khuey** [comment]
The idea being that if two files have identical extents then we can skip hashing?

**Comment by rocallahan** [comment]
Yeah. I've implemented this and a couple of other ideas. Files with the same extents (due to cloning/hardlinks) are all known to be the same so we can collapse them all to the same file without any more work. Also, if, after that collapsing, a file has a size that's different from all other files, we don't need to hash that either, we know it's unique. This should reduce the required hashes of mmap_clone_*_libxul.so in Karl's example from 88 to 0.

**Comment by rocallahan** [comment]
I need to write some automated tests for `rr pack` though before I land this.

**Comment by rocallahan** [comment]
Ok, `rr pack` should be infinitely faster now. I tested packing a Firefox trace on btrfs and it worked, and I tried to be paranoid, but I'm a little bit worried about regressions.

**Comment by karlt** [comment]
Confirming that `rr pack` now runs in 4 seconds with 132 libxul mappings, thank you!  Tested with successful `rr replay -a`.

FWIW I had ran similar steps with debug and optimized Firefox builds to confirm that something specific to the optimized build is triggering the additional mmaps.
With the optimized build, this run had 176 libxul mappings and rr pack took 69 minutes.  With the debug build, there were 20 libxul mappings and rr pack took no more than 8 minutes.
That's irrelevant now though.  Thanks!
